File: host_core/input_data.cpp
#include "input_data.h"

#include <getopt.h>

using namespace std;


/// Init static variable
Input_data* Input_data::_instance = nullptr;

Input_data::Input_data ( int argc, char* argv[] ) :
_dbg        ( "#log: Input_data - " ),
_verbose    ( 0 ),
_timer      ( 0 ),
_in_file    ( "" ),
_out_file   ( "" ),
_store_type ( 0 ) {
  
  if ( argc < 2 ) {
    print_help();
    exit ( 0 );
  }

  /// Read input
  int c;
  int verbose_flag = 0, time_flag = 0;
  while ( true ) {
    static struct option long_options[] =
    {
      /// These options set a flag
      {"verbose",       no_argument, &_verbose,       1}, /// Print verbose during computation
      {"time",          no_argument, &_timer,         1}, /// Print computational time
      /* 
       * These options don't set a flag.
       * We distinguish them by their indices.
       */
      {"help",          no_argument,       0,       'h'}, /// Print a help message
      {"input",         required_argument, 0,       'i'}, /// Set input file
      {"output",        required_argument, 0,       'o'}, /// Set output file
      {"store",         required_argument, 0,       's'}, /// Select store type
      {0, 0, 0, 0}
    };
    
    /// getopt_long stores the option index here
    int option_index = 0;
    c = getopt_long (argc, argv, "vthi:o:s:",
                     long_options, &option_index);
    /// Detect the end of the options
    if ( c == -1 ) break;
    /// Switch on c to detect the input given by the user
    switch ( c ) {
      case 0:
        /// If this option set a flag, do nothing else now
        if ( long_options[ option_index ].flag != 0 )
          break;
        printf ( "option %s", long_options[ option_index ].name );
        if ( optarg )
          printf (" with arg %s", optarg);
        printf ("\n");
        break;
      case 'h':
        print_help();
        exit( 0 );
      case 'v':
        _verbose = true;
        break;
      case 'i':
        _in_file  = optarg;
        break;
      case 'o':
        _out_file = optarg;
        break;
      case 't':
        _timer = true;
        break;
      case 's':
        _store_type = atoi( optarg );
        break;
      default:
        exit( 0 );
    }//switch
  }//while
  
  /// Exit if the user did not set an input file!
  if ( _in_file == "" ) {
    print_help();
    exit( 0 );
  }
  if ( _verbose   ) puts ("verbose flag is set");
  
  /* Print any remaining command line arguments (not options). */
  if ( optind < argc ) {
    printf ("non-option ARGV-elements: ");
    while ( optind < argc )
      printf ("%s ", argv[ optind++ ]);
    putchar ('\n');
  }
}//Input_data

string
Input_data::get_ifile () const {
  return _in_file;
}//get_ifile

string
Input_data::get_ofile () const {
  return _out_file;
}//get_ifile

bool
Input_data::timer_on () const {
  return _timer;
}//get_ifile

bool
Input_data::verbose () const {
  return _verbose;
}//get_ifile

//Factory_store::DataStoreType
int
Input_data::get_store_type () {
  return _store_type;
}//DataStoreType

void
Input_data::print_help () {
  string spaces = "        ";
  cout << "Usage: ./concordance -i <infile> [options]\n" << endl;
  cout << "          Options           |          Description      \n";
  cout << "=========================== | ===========================\n";
  cout << " -v|--verbose               | - Printf verbose info\n";
  cout << "                            |   during computation.\n";
  cout << " -t|--time                  | - Print computational time.\n";
  cout << " -h|--help                  | - Print this help message.\n";
  cout << " -i|--input      (string)   | - Read input file.\n";
  cout << " -o|--output     (string)   | - Set output file.\n";
  cout << "                            |   Default: text console.\n";
  cout << " -s|--store      (int)      | - Set type of store to use:\n";
  cout << "                            |   0 - Prefix tree.\n";
  cout << "                            |   1 - Red-Black tree.\n";
  cout << "                            |   Default: Prefix tree.\n";
  cout << "=========================== | ===========================\n";
  cout << "You may want to try:\n";
  cout << "\t" << "./concordance -i test/test_input.txt\n";
  cout << "For any questions, feel free to write at: campe8@nmsu.edu.\n";
}//print_help

File: Concordance_Java/src/concordance/Input_data.java
/****************************************************
 * Input data class:                                *
 * Singleton class used for reading                 *
 * input options given by the user.                 *
 ****************************************************/

package concordance;

class Input_data {
  /* Debug string */
  private String dbg;
  /* Static instance for singleton obj */
  protected static Input_data instance = null;
  /* General info */
  private boolean verbose;
  private boolean timer;
  private int     storeType;
  private String  iFile;
  private String  oFile;
  
  private void printHelp() {
    String helpString = 
              "          Options           |          Description      \n"
            + "=========================== | ===========================\n"
            + " -v|--verbose               | - Printf verbose info\n"
            + "                            |   during computation.\n"
            + " -t|--time                  | - Print computational time.\n"
            + " -h|--help                  | - Print this help message.\n"
            + " -i|--input      (string)   | - Read input file.\n"
            + " -o|--output     (string)   | - Set output file.\n"
            + "                            |   Default: text console.\n"
            + " -s|--store      (int)      | - Set type of store to use:\n"
            + "                            |   0 - Prefix tree.\n"
            + "                            |   1 - Red-Black tree.\n"
            + "                            |   Default: Prefix tree.\n"
            + "=========================== | ===========================\n"
            + "You may want to try:\n"
            + "\tjava -jar Concordance.jar -i test/test_input.txt\n"
            + "For any questions, feel free to write at: campe8@nmsu.edu.\n";
    System.out.println( "Usage: java -jar Concordance.jar -i <infile> [options]" );
    System.out.println( helpString );
  }//printHelp
  
  /* Protected constructor: a client that tries to instantiate
   * Singleton directly will get an error.*/ 
  protected Input_data( String[] args ) {
    init();
    if ( args.length < 1 ) {
      printHelp();
      System.exit( 0 );
    }
    else {
      /* Parsing the inputs... */
      for ( int i = 0; i < args.length; i++ ) {
        if ( args[ i ].contentEquals( "--verbose" ) || 
             args[ i ].contentEquals( "-v" ) ) {
          verbose = true; 
          continue;
        }
        if ( args[ i ].contentEquals( "--time" ) || 
             args[ i ].contentEquals( "-t" ) ) {
          timer = true; 
          continue;
        }
        if ( args[ i ].contentEquals( "--input" ) || 
             args[ i ].contentEquals( "-i" ) ) {
          if ( (i+1) == args.length ) break;
          iFile = args[ i + 1 ]; 
          continue;
        }
        if ( args[ i ].contentEquals( "--output" ) || 
             args[ i ].contentEquals( "-o" ) ) {
          if ( (i+1) == args.length ) break;
          oFile = args[ i + 1 ]; 
          continue;
        }
        if ( args[ i ].contentEquals( "--store" ) || 
             args[ i ].contentEquals( "-s" ) ) {
          if ( (i+1) == args.length ) break;
          try {
            storeType = Integer.parseInt( args[ i + 1 ] );
          } catch ( NumberFormatException nfe ) { storeType = 0; }
          continue;
        }
        if ( args[ i ].contentEquals( "--help" ) || 
             args[ i ].contentEquals( "-h" ) ) {
          printHelp();
          System.exit( 0 );
        }
      }
    }
  }//Input_data
  
  private void init() {
    dbg = "#log: Input_data - ";
    verbose   = false;
    timer     = false;
    iFile     = "";
    oFile     = "";
    storeType = 0;
  }//init
  
  /* Constructor:
   * Get the unique instance (static) of singleton object. */
  public static Input_data getInstance( String[] args ) {
    if ( instance == null ) {
      instance = new Input_data( args );
    }
    return instance;
  }//getInstance
  
  /* Get Methods */
  public String  getIFile() { return iFile;     }
  public String  getOFile() { return oFile;     }
  public boolean timerOn()  { return timer;     }
  public boolean verbose()  { return verbose;   }
  public int getStoreType() { return storeType; }
}//Input_data
